runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+package duckdbreplicator
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log/slog"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/XSAM/otelsql"
+	"github.com/google/uuid"
+	"github.com/jmoiron/sqlx"
+	"github.com/marcboeker/go-duckdb"
+	"github.com/mitchellh/mapstructure"
+	"go.opentelemetry.io/otel/attribute"
+	"gocloud.dev/blob"
+)
+
+type DB interface {
+	// Close closes the database.
+	Close() error
+
+	// AcquireReadConnection returns a connection to the database for reading.
+	// Once done the connection should be released by calling the release function.
+	// This connection must only be used for select queries or for creating and working with temporary tables.
+	AcquireReadConnection(ctx context.Context) (conn Conn, release func() error, err error)
+
+	// AcquireWriteConnection returns a connection to the database for writing.
+	// Once done the connection should be released by calling the release function.
+	// Any persistent changes to the database should be done by calling CRUD APIs on this connection.
+	AcquireWriteConnection(ctx context.Context) (conn Conn, release func() error, err error)
+
+	// Size returns the size of the database in bytes.
+	// It is currently implemented as sum of the size of all serving `.db` files.
+	Size() int64
+
+	// CRUD APIs
+
+	// CreateTableAsSelect creates a new table by name from the results of the given SQL query.
+	CreateTableAsSelect(ctx context.Context, name string, sql string, opts *CreateTableOptions) error
+
+	// InsertTableAsSelect inserts the results of the given SQL query into the table.
+	InsertTableAsSelect(ctx context.Context, name string, sql string, opts *InsertTableOptions) error
+
+	// DropTable removes a table from the database.
+	DropTable(ctx context.Context, name string) error
+
+	// RenameTable renames a table in the database.
+	RenameTable(ctx context.Context, oldName, newName string) error
+
+	// AddTableColumn adds a column to the table.
+	AddTableColumn(ctx context.Context, tableName, columnName, typ string) error
+
+	// AlterTableColumn alters the type of a column in the table.
+	AlterTableColumn(ctx context.Context, tableName, columnName, newType string) error
+}
+
+type DBOptions struct {
+	// Clean specifies whether to start with a clean database or download data from cloud storage and start with backed up data.
+	Clean bool
```

Comment:
That can be problematic right. If we can not create handle because db files are corrupted or can't be attached for whatever reason then `NewDB` will not return an initialized database which means clients can not call `Clean()` API.

Reply:
So this was specifically to handle this config : https://github.com/rilldata/rill/blob/e9a8c6c1966084962771546ab8ace5af7db8c500/runtime/drivers/duckdb/config.go#L33-L34

Basically which meant we do not cleanup file when this is set to true but return error. But given duckdb is now backward compatible, I will go ahead and remove this config and `Clean` option.